### PR TITLE
Backend data API endpoints

### LINKED ISSUE #50: Develop API endpoints for AcceptedCategory

### PR DESCRIPTION
**Pull request by malcolmrjones**
Fixes  #50  and #51 

This PR adds initial support for the data backend endpoints. Not all planned endpoints are supported but the following endpoints are supported.

-  /data/categories
- /data/acceptedcategory/:id
- /data/acceptedcategory/organization/:id
- /data/acceptedcategory/category/:id

Testing is done using the [firestore-mock](https://www.npmjs.com/package/firestore-mock) package. Due to the limitations and different behavior of mock Firestore not all endpoints can be covered. The following endpoints cannot and are not covered by tests.

- /data/acceptedcategory/organization/:id
- /data/acceptedcategory/category/:id

### CODE REVIEW

File: routes/data.js [hunk 1]
```diff
@@ -1,57 +1,127 @@
-const {Firestore} = require('@google-cloud/firestore');
+const { Firestore } = require('@google-cloud/firestore');
```
Comment (SabrinaSnider) [inline routes/data.js:1]: Nit: spacing
`const {Firestore}`

File: routes/data.js [hunk 2]
```diff
@@ -1,57 +1,127 @@
...
+let firestore = new Firestore();
+
+/**
+ * Utility function that returns the correct collection name depending on whther the environment is PRODUCTION or DEVELOPMENT.
+ * @param {String} collectionName
+ */
+function resolveCollectionName(collectionName) {
+  return process.env.NODE_ENV == 'production'
+    ? collectionName
+    : 'dev-' + collectionName;
```
Comment (SabrinaSnider) [inline routes/data.js:13]: Did you space this out for clarity?

File: test/test-data.js [hunk 3]
```diff
@@ -0,0 +1,215 @@
+const expect = require('chai').expect;
+const assert = require('assert');
+const request = require('request');
+const data = require('../routes/data');
+const FirestoreMock = require('firestore-mock');
+const { Firestore } = require('@google-cloud/firestore');
```
Comment (SabrinaSnider) [inline test/test-data.js:6]: Nit: Spacing
`const {Firestore}`

File: routes/data.js [hunk 4]
```diff
@@ -1,52 +1,127 @@
...
-    res.send(data);
-  });
-};
+let firestore = new Firestore();
 
 /**
- * Inserts 3 new documents into the 'movies' collection with hard-coded data.
+ * Utility function that returns the correct collection name depending on whther the environment is PRODUCTION or DEVELOPMENT.
+ * @param {String} collectionName
  */
```
Comment (danielabreo) [inline routes/data.js:9]: Add return documentation.

File: routes/data.js [hunk 5]
```diff
@@ -1,52 +1,127 @@
...
-      title: 'Iron Man',
-      year: 2008,
-    },
-  ];
-
-  movieData.forEach((movie) => {
-    const document = firestore.collection(DB_COLLECTION_NAME).doc();
-    document.set(movie);
-  });
+function resolveCollectionName(collectionName) {
```
Comment (danielabreo) [inline routes/data.js:10]: nit: an alternative for this function is to put the `collectionName` in the .env file so changes automatically based on the command.

File: routes/data.js [hunk 6]
```diff
@@ -1,52 +1,127 @@
...
+  firestore = firestoreInstance;
+};
+
+exports.getDatabase = function (firestoreInstance) {
+  return firestore;
+};
+
 /**
- * Read all documents in the 'movies' collection
+ * Queries for all AcceptedCategories by either organization or category depending on the  document reference and fieldName
```
Comment (danielabreo) [inline routes/data.js:25]: nit: remove extra space
```suggestion
 * Queries for all AcceptedCategories by either organization or category depending on the document reference and fieldName
```

File: routes/data.js [hunk 7]
```diff
@@ -1,52 +1,127 @@
...
+ * @param {FirebaseFirestore.DocumentReference} ref
+ * @param {String} fieldName
  */
```
Comment (danielabreo) [inline routes/data.js:28]: nit: add return type documentation

File: routes/data.js (hunk 2 above)
Comment (malcolmrjones) [reply routes/data.js]: That was formatted automatically by my local prettier but the formatting has changed now.

File: routes/data.js [hunk 8]
```diff
@@ -1,52 +1,120 @@
...
 }
 
+exports.setDatabase = function (firestoreInstance) {
+  firestore = firestoreInstance;
+};
+
+exports.getDatabase = function (firestoreInstance) {
+  return firestore;
+};
+
```
Comment (brucezheng) [inline routes/data.js:22]: You can use lambda notation e.g. ```firestoreInstance => firestore```.

File: routes/data.js [hunk 9]
```diff
@@ -1,52 +1,120 @@
...
+    .get();
+  const results = {};
+  snapshot.docs.forEach((doc) => {
+    results[doc.id] = doc.data();
   });
-  return movieList;
+  return results;
 }
+
+/** ****** Response Handlers **************/
```
Comment (brucezheng) [inline routes/data.js:41]: Just do

```/* Response Handlers */```

File: routes/data.js [hunk 10]
```diff
@@ -1,52 +1,120 @@
...
-  });
+function resolveCollectionName(collectionName) {
+  return process.env.NODE_ENV == 'production' ? collectionName : 'dev-' + collectionName;
 }
 
+exports.setDatabase = function (firestoreInstance) {
+  firestore = firestoreInstance;
+};
+
+exports.getDatabase = function (firestoreInstance) {
```
Comment (brucezheng) [inline routes/data.js:19]: Why does this have an argument?

File: test/test-data.js [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    function (error, response, body) {
+      expect(response.statusCode).to.equal(404);
+      done();
+    }
+  );
+});
+
+it('POST Request /data/acceptedcategories/organization/:id : add new AcceptedCategory', function (done) {
+  const oldCollectionLength = Object.keys(firestoreMock._db._collections['dev-AcceptedCategories'])
+    .length;
```
Comment (brucezheng) [inline test/test-data.js:169]: Nit:

It's cleaner if you break like

```
const oldCollectionLength =
    Object.keys(firestoreMock._db._collections['dev-AcceptedCategories']).length;
```

File: routes/data.js (hunk 10 above)
Comment (malcolmrjones) [reply routes/data.js:19]: Copy and paste error

File: test/test-data.js (hunk 11 above)
Comment (malcolmrjones) [reply test/test-data.js:166]: This line was automatically formatted. If I change the formatting to your suggestion will will fail the **Code Quality Check / quality-check (pull_request)** check.

**Pull request merged**